src/Psalm/Internal/Analyzer/Statements/Expression/BinaryOp/ArithmeticOpAnalyzer.php

```diff
@@ -867,4 +895,454 @@ public static function arithmeticOperation(
+            $min_operand2 = $right_type_part->max_bound;
+            //and the max is the max from first range minus the min from the second
+            $max_operand1 = $left_type_part->max_bound;
+            $max_operand2 = $right_type_part->min_bound;
+        } else {
+            $min_operand1 = $left_type_part->min_bound;
+            $min_operand2 = $right_type_part->min_bound;
+
+            $max_operand1 = $left_type_part->max_bound;
+            $max_operand2 = $right_type_part->max_bound;
+        }
+
+        $calculated_min_type = null;
+        if ($min_operand1 !== null && $min_operand2 !== null) {
+            // when there are two valid numbers, make any operation
+            $calculated_min_type = self::arithmeticOperation(
+                $parent,
+                $min_operand1,
+                $min_operand2,
+                false
+            );
+        }
+
+        $calculated_max_type = null;
+        if ($max_operand1 !== null && $max_operand2 !== null) {
+            // when there are two valid numbers, make any operation
+            $calculated_max_type = self::arithmeticOperation(
+                $parent,
+                $max_operand1,
+                $max_operand2,
+                false
+            );
+        }
+
+        $min_value = $calculated_min_type !== null ? $calculated_min_type->getSingleIntLiteral()->value : null;
+        $max_value = $calculated_max_type !== null ? $calculated_max_type->getSingleIntLiteral()->value : null;
+
+        $new_result_type = new Type\Union([new Type\Atomic\TIntRange($min_value, $max_value)]);
+
+        if (!$result_type) {
+            $result_type = $new_result_type;
+        } else {
+            $result_type = Type::combineUnionTypes($new_result_type, $result_type);
+        }
+    }
+
+    /**
+     * @param TIntRange|TInt $left_type_part
+     * @param TIntRange|TInt $right_type_part
+     */
+    private static function analyzeOperandsBetweenIntRangeAndInt(
+        PhpParser\Node $parent,
+        ?Type\Union &$result_type,
+        Atomic $left_type_part,
+        Atomic $right_type_part
+    ): void {
+        if ($left_type_part instanceof Type\Atomic\TIntRange) {
+            $left_is_range = true;
+            $range_operand = $left_type_part;
+            $other_operand = $right_type_part;
+        } elseif ($right_type_part instanceof Type\Atomic\TIntRange) {
+            $left_is_range = false;
+            $range_operand = $right_type_part;
+            $other_operand = $left_type_part;
+        } else {
+            //this can't happen
+            return;
+        }
+
+        $new_range = TIntRange::convertToIntRange($other_operand);
+        if ($left_is_range) {
+            self::analyzeOperandsBetweenIntRange($parent, $result_type, $range_operand, $new_range);
+        } else {
+            self::analyzeOperandsBetweenIntRange($parent, $result_type, $new_range, $range_operand);
+        }
```

Comment:
```suggestion
        if (!$left_type_part instanceof Type\Atomic\TIntRange) {
            $left_type_part = TIntRange::convertToIntRange($left_type_part);
        }
        if (!$right_type_part instanceof Type\Atomic\TIntRange) {
            $right_type_part = TIntRange::convertToIntRange($right_type_part);
        }
        return self::analyzeOperandsBetweenIntRange($parent, $result_type, $left_type_part, $right_type_part);
```

Reply:
Wow, I wrote so many line for so little :D Thanks!